src/base64.cpp

```diff
@@ -110,14 +110,23 @@ static_assert( length(translation_table) == 64, "error: need 64 characters in th
 // simple encoding function for 3 bytes
 void encode( const std::uint8_t*& byte_array, std::string::iterator& writer )
 {
-	const uint32_t* bits = reinterpret_cast<const uint32_t*>(byte_array);
-	for(int i = 0; i < 4; ++i)
-	{
-		unsigned index = bitmask & (*bits >> (6*i));
-		assert(index < 64);
-		*writer = translation_table[index];
-		++writer;
-	}
+	unsigned const index0 = bitmask & (byte_array[0] >> 2);
+	assert(index0 < 64);
+	writer[0] = translation_table[index0];
+
+	unsigned const index1 = bitmask & ((byte_array[0] << 4) | (byte_array[1] >> 4));
+	assert(index1 < 64);
+	writer[1] = translation_table[index1];
+
+	unsigned const index2 = bitmask & ((byte_array[1] << 2) | (byte_array[2] >> 6));
+	assert(index2 < 64);
+	writer[2] = translation_table[index2];
+
+	unsigned const index3 = bitmask & (byte_array[2]);
+	assert(index3 < 64);
+	writer[3] = translation_table[index3];
+
+	writer += 4;
```

Comment:
shouldn't this also use `std::advance` to be consistent? Also, given that we use an iterator interface here, should we use `writer[i]`? Or actually consider this to be an `output_iterator` and perform iterator increments and writes?

Reply:
I changed to `std::advance`. In my opinion using `writer[i]` is ok here.

Reply:
ok